tests/utils.py

```diff
@@ -125,15 +119,22 @@ def __call__(self, estimator):
         if isinstance(estimator, (LinearClassifierMixin, SVC, NuSVC,
                                   LightBaseClassifier)):
             y_pred = estimator.decision_function(self.X_test)
-        elif isinstance(estimator, DecisionTreeClassifier):
-            y_pred = estimator.predict_proba(self.X_test.astype(np.float32))
         elif isinstance(
                 estimator,
-                (ForestClassifier, XGBClassifier, LGBMClassifier)):
+                (ForestClassifier, DecisionTreeClassifier,
+                 XGBClassifier, LGBMClassifier)):
             y_pred = estimator.predict_proba(self.X_test)
         else:
             y_pred = estimator.predict(self.X_test)
 
+        # Some models force input data to be particular type
+        # during prediction phase in their native Python libraries.
+        # For correct comparison of testing results we mimic the same behavior
+        if isinstance(estimator, (BaseDecisionTree, BaseForest)):
+            self.X_test = self.X_test.astype(np.float32, copy=False)
```

Comment:
Interestingly that XGBoost also cast all inputs to `float` during `predict`.

https://github.com/dmlc/xgboost/blob/12110c900eff0aaa06045ecf717e6c5a36a164d5/python-package/xgboost/sklearn.py#L595-L596

https://github.com/dmlc/xgboost/blob/12110c900eff0aaa06045ecf717e6c5a36a164d5/python-package/xgboost/core.py#L433-L438

https://github.com/dmlc/xgboost/blob/12110c900eff0aaa06045ecf717e6c5a36a164d5/python-package/xgboost/data.py#L493-L495

https://github.com/dmlc/xgboost/blob/12110c900eff0aaa06045ecf717e6c5a36a164d5/python-package/xgboost/data.py#L135

https://github.com/dmlc/xgboost/blob/12110c900eff0aaa06045ecf717e6c5a36a164d5/python-package/xgboost/data.py#L115-L116

https://github.com/dmlc/xgboost/blob/12110c900eff0aaa06045ecf717e6c5a36a164d5/python-package/xgboost/data.py#L138

But including it here makes a lot of tests to fail...

Reply:
I'm a little concerned that here we treat a symptom not the cause. By casting the input vector to float32 and them passing it as strings into estimators we don't exactly reproduce the actual environment, where  casted values will be transformed back to doubles because of the `score` function signature. What do you think? Am I overthinking this?

Reply:
> I'm a little concerned that here we treat a symptom not the cause.

Yeah, you're absolutely right! To fix the root cause we should support multiple floating types in target languages. In this PR I just propose to make our tests a little bit fairer. Native libraries do `double -> float` conversion and in our tests we perform `double -> float -> double`. Unfortunately, I'm not a numerical expert but it seems that `float -> double`  is a safe conversion: https://stackoverflow.com/questions/29648271/convert-float-double-float.